batch_job/src/batch_job_mesos.c

```diff
@@ -0,0 +1,422 @@
+		}
+		fprintf(task_info_fp, "%s,", (mt->task_input_files->items)[num_input_files-1]);
+
+	} else {
+		fputs(",", task_info_fp);
+	}
+    
+	if (extra_output_files != NULL && strlen(extra_output_files) != 0) {
+		int j = 0;
+		int num_output_files = mt->task_output_files->used_length;
+		for (j = 0; j < (num_output_files-1); j++) {
+			fprintf(task_info_fp, "%s ", (mt->task_output_files->items)[j]);
+		}
+		fprintf(task_info_fp, "%s,", (mt->task_output_files->items)[num_output_files-1]);
+	} else {
+		fputs(",", task_info_fp);
+	}
+
+	int64_t cores = 4;
+	int64_t memory = 5120;
+	int64_t disk = 5120;
+
+	if (resources) {
+		cores  = resources->cores  > -1 ? resources->cores  : cores;
+		memory = resources->memory > -1 ? resources->memory : memory;
+		disk   = resources->disk   > -1 ? resources->disk   : disk;
+	}
+
+	fprintf(task_info_fp, "%" PRId64 ",%" PRId64 ",%" PRId64 ",", cores, memory, disk);
+
+	fputs("submitted\n", task_info_fp);
+
+	mesos_task_delete(mt);
+	fclose(task_info_fp);
+
+	return task_id;
+}
+
+static batch_job_id_t batch_job_mesos_wait (struct batch_queue * q, struct batch_job_info * info_out, time_t stoptime)
+{
+		
+	char *line = NULL;
+	size_t len = 0;
+	ssize_t read_len;
+	FILE *task_state_fp;
+
+	if(finished_tasks == NULL) {
+		finished_tasks = itable_create(0);
+	}
+
+	while(access(FILE_TASK_STATE, F_OK) == -1) {}
+
+	task_state_fp = fopen(FILE_TASK_STATE, "r");
+
+	while(1) {
+
+		char *task_id_ch;
+		char *task_stat_str;
+		int task_id;
+				
+		while((read_len = getline(&line, &len, task_state_fp)) != -1) {
+
+			// trim the newline character
+			if (line[read_len-1] == '\n') {
+				line[read_len-1] = '\0';
+				--read_len;
+			}
+
+			task_id_ch = strtok(line, ",");
+			task_id = atoi(task_id_ch);
+
+			// There is a new task finished
+			if(itable_lookup(finished_tasks, task_id) == NULL) {
+				struct batch_job_info *info = itable_remove(q->job_table, task_id);
+			    	
+				info->finished = time(0);
+				task_stat_str = strtok(NULL, ",");
+
+				if (strcmp(task_stat_str, "finished") == 0) {
+					info->exited_normally = 1;
+				} else {
+					info->exited_normally = 0;
+				}
+
+				memcpy(info_out, info, sizeof(*info));
+				free(info);
+				fclose(task_state_fp);
+
+				int itable_val = 1;
+				itable_insert(finished_tasks, task_id, &itable_val);
+
+				return task_id;
+			}
+		}
+		sleep(1);
+
+		if(stoptime != 0 && time(0) >= stoptime) {
+			fclose(task_state_fp);
+			return -1;
+		}
+	}
+
+}
+
+static int batch_job_mesos_remove (struct batch_queue *q, batch_job_id_t jobid)
+{
+	struct batch_job_info *info = itable_lookup(q->job_table, jobid);
+	info->finished = time(0);
+	info->exited_normally = 0;
+	info->exit_signal = 0;
+	// append the new task state to the "mesos_task_info" file
+	char *cmd = string_format("awk -F \',\' \'{$(NF--)=\"aborting\"; if($1==\"%" PRIbjid "\"){for(i=1;i<NF;i+=1){printf $i\",\"}; print $NF}}\' %s >> %s", jobid, FILE_TASK_INFO, FILE_TASK_INFO);
```

Comment:
Wow, awk is a rather expensive way to go about this.
Could you do this with direct file access instead?